Very long creation time when trying to create an already existing type

```
module.ct-dev.commercetools_type.category: Creating...
2019/10/31 13:02:47 [DEBUG] module.ct-dev.commercetools_type.category: applying the planned Create change
2019/10/31 13:02:47 [TRACE] GRPCProvider: ApplyResourceChange
2019-10-31T13:02:47.181+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:47 [DEBUG] Waiting for state to become: [success]
2019-10-31T13:02:47.511+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:47 [TRACE] Waiting 500ms before next try
2019-10-31T13:02:48.047+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:48 [TRACE] Waiting 1s before next try
2019-10-31T13:02:49.081+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:49 [TRACE] Waiting 2s before next try
2019-10-31T13:02:51.126+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:51 [TRACE] Waiting 4s before next try
2019-10-31T13:02:55.172+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:02:55 [TRACE] Waiting 8s before next try
module.ct-dev.commercetools_type.category: Still creating... [10s elapsed]
2019-10-31T13:03:03.220+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:03 [TRACE] Waiting 10s before next try
module.ct-dev.commercetools_type.category: Still creating... [20s elapsed]
2019-10-31T13:03:13.266+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:13 [TRACE] Waiting 10s before next try
module.ct-dev.commercetools_type.category: Still creating... [30s elapsed]
2019-10-31T13:03:23.303+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:23 [TRACE] Waiting 10s before next try
module.ct-dev.commercetools_type.category: Still creating... [40s elapsed]
2019-10-31T13:03:33.357+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:33 [TRACE] Waiting 10s before next try
module.ct-dev.commercetools_type.category: Still creating... [50s elapsed]
2019-10-31T13:03:43.400+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:43 [TRACE] Waiting 10s before next try
module.ct-dev.commercetools_type.category: Still creating... [1m0s elapsed]
2019-10-31T13:03:47.181+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:47 [WARN] WaitForState timeout after 1m0s
2019-10-31T13:03:47.181+0100 [DEBUG] plugin.terraform-provider-commercetools_v0.19.0: 2019/10/31 13:03:47 [WARN] WaitForState starting 30s refresh grace period
2019/10/31 13:03:47 [DEBUG] module.ct-dev.commercetools_type.category: apply errored, but we're indicating that via the Error pointer rather than returning it: A duplicate value '"category"' exists for field 'key'.
2019/10/31 13:03:47 [TRACE] module.ct-dev: eval: *terraform.EvalMaybeTainted
2019/10/31 13:03:47 [TRACE] EvalMaybeTainted: module.ct-dev.commercetools_type.category encountered an error during creation, so it is now marked as tainted
2019/10/31 13:03:47 [ERROR] module.ct-dev: eval: *terraform.EvalApplyPost, err: A duplicate value '"category"' exists for field 'key'.
2019/10/31 13:03:47 [ERROR] module.ct-dev: eval: *terraform.EvalSequence, err: A duplicate value '"category"' exists for field 'key'.
2019/10/31 13:03:47 [TRACE] [walkApply] Exiting eval tree: module.ct-dev.commercetools_type.category

Error: A duplicate value '"category"' exists for field 'key'.

  on terraform/ct-dev/main.tf line 1, in resource "commercetools_type" "category":
   1: resource "commercetools_type" "category" {
```

### What do you expect?

Since my tf state is empty and the type already exists the command should immediately exit with `type with key "foo" already exists! Remove it from commercetool or import the state`

--- a/commercetools/resource_api_client.go
+++ b/commercetools/resource_api_client.go
@@ -3,7 +3,9 @@ package commercetools
 import (
 	"sort"
 	"strings"
+	"time"
 
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/labd/commercetools-go-sdk/commercetools"
 )
@@ -50,7 +52,19 @@ func resourceAPIClientCreate(d *schema.ResourceData, m interface{}) error {
 	}
 
 	client := getClient(m)
-	apiClient, err := client.APIClientCreate(draft)
+
+	var apiClient *commercetools.APIClient
+
+	err := resource.Retry(20*time.Second, func() *resource.RetryError {
+		var err error
+
+		apiClient, err = client.APIClientCreate(draft)
+		if err != nil {
+			return handleCommercetoolsError(err)
+		}
+		return nil
+	})
+
 	if err != nil {
 		return err
 	}

--- a/commercetools/resource_api_extension.go
+++ b/commercetools/resource_api_extension.go
@@ -130,9 +130,7 @@ func resourceAPIExtensionCreate(d *schema.ResourceData, m interface{}) error {
 
 		extension, err = client.ExtensionCreate(draft)
 		if err != nil {
-			log.Print("[DEBUG] Error while creating extension, will try again")
-			log.Print(err)
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_channel.go
+++ b/commercetools/resource_channel.go
@@ -1,6 +1,9 @@
 package commercetools
 
 import (
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/labd/commercetools-go-sdk/commercetools"
 )
@@ -59,7 +62,18 @@ func resourceChannelCreate(d *schema.ResourceData, m interface{}) error {
 	}
 
 	client := getClient(m)
-	channel, err := client.ChannelCreate(draft)
+	var channel *commercetools.Channel
+
+	err := resource.Retry(20*time.Second, func() *resource.RetryError {
+		var err error
+
+		channel, err = client.ChannelCreate(draft)
+		if err != nil {
+			return handleCommercetoolsError(err)
+		}
+		return nil
+	})
+
 	if err != nil {
 		return err
 	}

--- a/commercetools/resource_product_type.go
+++ b/commercetools/resource_product_type.go
@@ -212,14 +212,7 @@ func resourceProductTypeCreate(d *schema.ResourceData, m interface{}) error {
 
 		ctType, err = client.ProductTypeCreate(draft)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_shipping_method.go
+++ b/commercetools/resource_shipping_method.go
@@ -73,14 +73,7 @@ func resourceShippingMethodCreate(d *schema.ResourceData, m interface{}) error {
 
 		shippingMethod, err = client.ShippingMethodCreate(draft)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_shipping_zone.go
+++ b/commercetools/resource_shipping_zone.go
@@ -77,7 +77,7 @@ func resourceShippingZoneCreate(d *schema.ResourceData, m interface{}) error {
 
 		shippingZone, err = client.ZoneCreate(draft)
 		if err != nil {
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_shipping_zone_rate.go
+++ b/commercetools/resource_shipping_zone_rate.go
@@ -160,16 +160,10 @@ func resourceShippingZoneRateCreate(d *schema.ResourceData, m interface{}) error
 
 	err = resource.Retry(1*time.Minute, func() *resource.RetryError {
 		var err error
+
 		shippingMethod, err = client.ShippingMethodUpdateWithID(&input)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_state.go
+++ b/commercetools/resource_state.go
@@ -1,6 +1,9 @@
 package commercetools
 
 import (
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
 	"github.com/labd/commercetools-go-sdk/commercetools"
@@ -102,7 +105,18 @@ func resourceStateCreate(d *schema.ResourceData, m interface{}) error {
 	}
 
 	client := getClient(m)
-	state, err := client.StateCreate(draft)
+	var state *commercetools.State
+
+	err := resource.Retry(20*time.Second, func() *resource.RetryError {
+		var err error
+
+		state, err = client.StateCreate(draft)
+		if err != nil {
+			return handleCommercetoolsError(err)
+		}
+		return nil
+	})
+
 	if err != nil {
 		return err
 	}

--- a/commercetools/resource_store.go
+++ b/commercetools/resource_store.go
@@ -1,6 +1,9 @@
 package commercetools
 
 import (
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/labd/commercetools-go-sdk/commercetools"
 )
@@ -41,7 +44,19 @@ func resourceStoreCreate(d *schema.ResourceData, m interface{}) error {
 	}
 
 	client := getClient(m)
-	store, err := client.StoreCreate(draft)
+
+	var store *commercetools.Store
+
+	err := resource.Retry(20*time.Second, func() *resource.RetryError {
+		var err error
+
+		store, err = client.StoreCreate(draft)
+		if err != nil {
+			return handleCommercetoolsError(err)
+		}
+		return nil
+	})
+
 	if err != nil {
 		return err
 	}

--- a/commercetools/resource_subscription.go
+++ b/commercetools/resource_subscription.go
@@ -228,6 +228,7 @@ func resourceSubscriptionCreate(d *schema.ResourceData, m interface{}) error {
 
 		subscription, err = client.SubscriptionCreate(draft)
 		if err != nil {
+			// Some subscription resources might not be ready yet, always keep retrying
 			return resource.RetryableError(err)
 		}
 		return nil

--- a/commercetools/resource_tax_category.go
+++ b/commercetools/resource_tax_category.go
@@ -65,14 +65,7 @@ func resourceTaxCategoryCreate(d *schema.ResourceData, m interface{}) error {
 
 		taxCategory, err = client.TaxCategoryCreate(draft)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_tax_category_rate.go
+++ b/commercetools/resource_tax_category_rate.go
@@ -144,14 +144,7 @@ func resourceTaxCategoryRateCreate(d *schema.ResourceData, m interface{}) error 
 	err = resource.Retry(30*time.Second, func() *resource.RetryError {
 		taxCategory, err = client.TaxCategoryUpdateWithID(input)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/resource_type.go
+++ b/commercetools/resource_type.go
@@ -238,14 +238,7 @@ func resourceTypeCreate(d *schema.ResourceData, m interface{}) error {
 
 		ctType, err = client.TypeCreate(draft)
 		if err != nil {
-			if ctErr, ok := err.(commercetools.ErrorResponse); ok {
-				if _, ok := ctErr.Errors[0].(commercetools.InvalidJSONInputError); ok {
-					return resource.NonRetryableError(ctErr)
-				}
-			} else {
-				log.Printf("[DEBUG] Received error: %s", err)
-			}
-			return resource.RetryableError(err)
+			return handleCommercetoolsError(err)
 		}
 		return nil
 	})

--- a/commercetools/utils.go
+++ b/commercetools/utils.go
@@ -3,8 +3,10 @@ package commercetools
 import (
 	"encoding/json"
 	"fmt"
+	"log"
 	"strings"
 
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/labd/commercetools-go-sdk/commercetools"
 )
@@ -17,6 +19,15 @@ const TypeLocalizedString = schema.TypeMap
 func getClient(m interface{}) *commercetools.Client {
 	client := m.(*commercetools.Client)
 	return client
+}
+
+func handleCommercetoolsError(err error) *resource.RetryError {
+	if ctErr, ok := err.(commercetools.ErrorResponse); ok {
+		return resource.NonRetryableError(ctErr)
+	}
+
+	log.Printf("[DEBUG] Received error: %s", err)
+	return resource.RetryableError(err)
 }
 
 func expandStringArray(input []interface{}) []string {

